Baikal/PostEffects/ML/denoiser.cpp

```diff
@@ -22,5 +52,150 @@ namespace Baikal
+
+            m_primitives->Normalize(0,
+                                    input_buf,
+                                    normalized_buf,
+                                    (int)input_buf.GetElementCount()).Wait();
+
+            m_context->ReadBuffer<Type>(0,
+                                        CLWBuffer<Type>::CreateFromClBuffer(
+                                        normalized_buf),
+                                        (Type*)(host_mem),
+                                        input_buf.GetElementCount()).Wait();
+        }
+
+        void MLDenoiser::Apply(InputSet const& input_set, Output& output)
+        {
+            auto shape = m_inference->GetInputShape();
+            auto width = std::get<0>(shape);
+            auto height = std::get<1>(shape);
+
+            auto tensor = m_inference->GetInputTensor();
+            auto host_mem = tensor.data();
+
+            for (const auto& input : input_set)
+            {
+                auto type = input.first;
+                auto clw_output = static_cast<ClwOutput*>(input.second);
+                auto device_mem = clw_output->data();
+
+                switch (type)
+                {
+                    case Renderer::OutputType::kColor:
+                    {
+                        auto mem_to_write = host_mem + m_layout[OutputType::kColor];
+                        ProcessOutput<cl_float3, RadeonRays::float3>(device_mem, mem_to_write);
+                        break;
+                    }
+                    case Renderer::OutputType::kDepth:
+                    {
+                        auto mem_to_write = host_mem + m_layout[OutputType::kDepth];
+                        ProcessOutput<cl_float, Tensor::ValueType>(device_mem, mem_to_write);
+                        break;
+                    }
+                    case Renderer::OutputType::kViewShadingNormal:
+                    {
+                        auto mem_to_write = host_mem + m_layout[OutputType::kViewShadingNormal];
+
+                        ProcessOutput<cl_float3, RadeonRays::float3>(device_mem,
+                            (Tensor::ValueType*)m_host_cache.get());
+
+                        // remove third chanel
+                        for (auto i = 0u; i < 3 * width * height; i += 3)
+                        {
+                            mem_to_write[i / 3] = m_host_cache[i];
+                            mem_to_write[i / 3 + 1] = m_host_cache[i + 1];
+                        }
+
+                        break;
+                    }
+                    case Renderer::OutputType::kGloss:
+                    {
+                        auto mem_to_write = host_mem + m_layout[OutputType::kGloss];
+                        
+                        ProcessOutput<cl_float3, RadeonRays::float3>(device_mem,
+                            (Tensor::ValueType*)m_host_cache.get());
+
+                        // remove second and third chanels
```

Comment:
it's better to write "copy only the first channel".

Reply:
Done